Enable Typedoc excludeNotExported

benmccann opened the pull request:
This turns on `excludeNotExported` in typedoc

It also contains a lot of workarounds for https://github.com/TypeStrong/typedoc/issues/393 to make it work

benmccann commented:
Sigh... This works locally, so I'm not sure why the tests are failing on the VMs

kurkle commented:
Probably need the exclude that was removed few days ago

benmccann commented:
What exclude?

kurkle commented:
Try if that one helps: https://github.com/chartjs/Chart.js/pull/7099#discussion_r379486483

kurkle commented:
Ok, no other ideas to try.

kurkle commented:
Oh, I read somewhere that you need to put the export on separate line in some cases (ie, end of file), when exporting a class as default (and using properties)

benmccann commented:
Well that would defeat the entire purpose of this PR :-p

benmccann commented:
This isn't worth spending more time on to me. I'd rather just not use `excludeNotExported`

kurkle commented:
I was supposed to comment here about moving `commonjs()` lower in the plugins. But it was `rollup-plugin-istanbul` that is failing here. I think its using babel 6 still.

I think I got it working in #7125 (need to extract the chores from that)

benmccann commented:
That worked! Thank you so much!!

benmccann commented:
Yeah, hopefully they fix that bug in Typedoc, but I think it's generally better this way anyway

kurkle commented on babel.config.json:
```diff
@@ -1,8 +1,8 @@
-    "@babel/plugin-transform-object-assign",
-    "@babel/plugin-proposal-class-properties"
+    "@babel/plugin-proposal-class-properties",
+    "@babel/plugin-transform-object-assign"
```
nit: order only changed

benmccann replied:
yeah, I had it this way in this PR, but one of the earlier ones got in first with a different order. I sort of preferred this order though because it's alphabetical

kurkle commented on karma.conf.js:
```diff
@@ -1,9 +1,9 @@
 /* eslint-disable import/no-commonjs */
 
+const babel = require('rollup-plugin-babel');
```
nit: order only changed

benmccann replied:
I left this change as well because it also alphabetizes it, but can revert if you don't think that's a good enough reason to change it

kurkle replied:
its good enough

kurkle commented on src/core/core.animator.js:
```diff
@@ -18,7 +18,7 @@ function drawFPS(chart, count, date, lastDate) {
 }
 
-class Animator {
+export class Animator {
```
Not sure about this, it enables going around the singleton pattern.

kurkle commented on src/core/core.defaults.js:
```diff
@@ -1,6 +1,6 @@
 import {merge} from '../helpers/helpers.core';
 
-class Defaults {
+export class Defaults {
```
Another singleton

kurkle commented on tsconfig.json:
```diff
@@ -10,6 +10,7 @@
   "typedocOptions": {
     "name": "Chart.js",
     "excludeExternals": true,
+    "excludeNotExported": true,
```
Does the directly exporting change anything else in the docs? Like not making those "external" or something.

benmccann replied:
No, I think the only change is that it hides some stuff

benmccann commented:
Ok, I just added some comments, but am just as happy to remove the `export` if you think that's preferrable

For the `Animator`, I think that class will mostly just be used internally and probably isn't something users would interact with, so removing the `export` and removing from the docs would be fine

For the `Defaults`, there's only one method: `set`. I just noticed that https://github.com/chartjs/Chart.js/pull/7074 removed the `_` prefix, so I just went ahead and removed the `@private`. I don't think the docs show this being used at all though and use alternate ways of setting the default. If you want to consider this class as internal for now as well I'm also happy to remove the `export`
